sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -57,7 +57,8 @@ case class BroadcastHashJoinExec(
   }
 
   override lazy val metrics = Map(
-    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+    "numMatchedRows" -> SQLMetrics.createMetric(sparkContext, "number of matched rows"))
```

Comment:
This seems like a "progress bar" instead of a metrics. Most likely people don't care about it because it's usually `num output rows of left` * `num output rows of right`.

Reply:
I am running the TPC-DS benchmark right now. I will post the results shortly.

Reply:
I ran the benchmark and I see a clear performance degradation for query-16 and query-94. All the other queries seem to be fine. I will debug why there is a performance degradation for  query-16 & query-94. https://docs.google.com/document/d/16_cjk3g3_7mMH2C_T9bsFroxaH_CihJkXINYirt-mtg/edit?usp=sharing 
cc: @cloud-fan

Reply:
Update: I figured out why the perf degradation is occurring for query-16 & query-94 and fixed it. I ran the benchmark for a couple of iterations and looks like my fix is working. I am running the benchmark with 25 iterations to identify if any other query has been effected.

Reply:
Update: In my most recent commit I fixed a perf issue that is causing a significant degradation for query-16 & quer-94. However there still seems to be some perf degradation that I cannot fix. Maybe it is because adding the new metric itself or maybe there is something wrong with the way that I am running the tests. I will run the benchmark again on a bare metal machine and see if it makes any difference. I will also see if there is anything else that can be done to improve the performance. My most recent runs are shared in this doc. https://docs.google.com/spreadsheets/d/1A0jCx5BL0wcN28oDQDOs73QhVZ71uNbKDg9Q8S0IE6g/edit?usp=sharing